### PR TITLE
Ensure spec tests actually run, and fix some failures

### PR DESCRIPTION
**Pull request by alexcrichton**
More details in the commits!

### CODE REVIEW

File: src/passes/used.rs [hunk 1]
```diff
@@ -64,6 +65,32 @@ impl Used {
             stack.push_func(f);
         }
 
+        // Initialization of imported memories or imported tables is a
```
Comment (alexcrichton) [inline src/passes/used.rs]: Note that this is arguably a bug in wasm-gc as-is today!

File: walrus-tests/tests/spec-tests.rs [hunk 2]
```diff
@@ -115,8 +115,21 @@ fn run(wast: &Path) -> Result<(), failure::Error> {
         .args(extra_args)
         .output()
         .context("executing `spectest-interp`")?;
+
+    // If the interpreter exits with success it may still have failed some
+    // tests. Check the output for `X/Y tests passed.` and make sure `X` equals
+    // `Y`.
     if output.status.success() {
```
Comment (fitzgen) [inline walrus-tests/tests/spec-tests.rs]: Seems like maybe we should file a bug upstream to exit with code = number of failures or something.

File: src/module/functions/local_function/context.rs [hunk 3]
```diff
@@ -143,8 +143,30 @@ impl<'a> FunctionContext<'a> {
...
         let expr = expr.into();
-        self.add_to_current_frame_block(expr);
-        impl_unreachable(&mut self.operands, &mut self.controls, expr);
+
+        let frame = self.controls.last_mut().unwrap();
+
+        // If we're not unreachable yet then we need to commit this expression.
+        // Note that we also need to commit any previous expressions on the
+        // stack because they may have side effects. For any lingering operands
+        // synthesize `Drop` nodes for them here.
```
Comment (fitzgen) [inline src/module/functions/local_function/context.rs]: Nice.

File: src/passes/used.rs (hunk 1 above)
Comment (fitzgen) [reply src/passes/used.rs]: > If a module imports a table/memory and has initializers for it, then we should consider that import live and retained to ensure initializers happen when the module is instantiated.

Yeah, that was an existing bug for sure

File: walrus-tests/tests/spec-tests.rs (hunk 2 above)
Comment (alexcrichton) [reply walrus-tests/tests/spec-tests.rs]: https://github.com/WebAssembly/wabt/issues/1002

**Pull request merged**